File: app/src/main/java/com/xaye/downloader/notify/DataChanger.java
package com.xaye.downloader.notify;


import android.content.Context;

import com.xaye.downloader.db.DownloadDatabase;
import com.xaye.downloader.entities.DownloadEntry;
import com.xaye.downloader.db.DownloadEntryDao;
import com.xaye.downloader.entities.DownloadStatus;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Observable;

/**
 * @FileName:com.xaye.downloader.DataChanger.kt Author xaye
 * @date: 2024-04-05 08:47
 * Created by 11623 on 2024/4/5
 */
public class DataChanger extends Observable {

    private static final String TAG = "DataChanger";
    private static DataChanger instance;

    private final Context mContext;
    private LinkedHashMap<String, DownloadEntry> m0peratedEntries;

    public static DataChanger getInstance(Context context) {
        if (instance == null) {
            instance = new DataChanger(context);
        }
        return instance;
    }

    private DataChanger(Context context) {
        this.mContext = context;
        m0peratedEntries = new LinkedHashMap<>();
    }

    public void postStatus(DownloadEntry entry) {

        m0peratedEntries.put(entry.getId(), entry);

        new Thread(() -> {

            DownloadEntryDao downloadDao = DownloadDatabase.getInstance(mContext).downloadEntryDao();
            DownloadEntry existingDownload = downloadDao.getDownloadById(entry.getId());

            //此时 entry 是ListActivity new 的 pid为 null 而 existingDownload 是数据库的 pid不为 null 的情况

            if (existingDownload != null) {
                //需要把pid 赋值给entry， 有了主键 数据库才会执行替换操作！
                entry.setPid(existingDownload.getPid());
                downloadDao.insertOrUpdate(entry);
                //Trace.d(TAG + " postStatus insertOrUpdate ->> existingDownload");
            } else {
                downloadDao.insertOrUpdate(entry);
                //Trace.d(TAG + " postStatus insertOrUpdate ->> entry");
            }

            List<DownloadEntry> datas = DownloadDatabase.getInstance(mContext).downloadEntryDao().getAllDownloads();

           // Trace.d(TAG + " postStatus datas size = " + datas.size() + " , datas = " + Arrays.toString(datas.toArray()));
        }).start();

        setChanged();
        notifyObservers(entry);
    }

    public ArrayList<DownloadEntry> queryAllRecoverableEntries() {
        ArrayList<DownloadEntry> entries = null;
        for (DownloadEntry entry : m0peratedEntries.values()) {
            if (entry.getStatus() == DownloadStatus.PAUSED) {
                if (entries == null) {
                    entries = new ArrayList<>();
                }
                entries.add(entry);
            }
        }

        return entries;
    }

    public DownloadEntry queryDownloadEntryById(String id) {
        return m0peratedEntries.get(id);
    }

    public void addToOperatedEntryMap(String id, DownloadEntry downloadEntry) {
        m0peratedEntries.put(id, downloadEntry);
    }

    public boolean constainsDownloadEntry(String id) {
        return m0peratedEntries.containsKey(id);
    }

    public boolean deleteDownloadEntry(String id) {
        DownloadDatabase.getInstance(mContext).downloadEntryDao().deleteDownloadById(id);
        return m0peratedEntries.remove(id) != null;
    }
}

File: app/src/main/java/com/xaye/downloader/notify/DataChanger.kt
package com.xaye.downloader.notify

import android.content.Context
import android.os.Handler
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.xaye.downloader.db.DownloadDatabase
import com.xaye.downloader.entities.DownloadEntry
import com.xaye.downloader.entities.DownloadStatus
import com.xaye.downloader.utilities.Trace
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

/**
 * Author xaye
 * @date: 2024-05-26 18:22
 */
class DataChanger private constructor(private val context: Context) {
    private val operatedEntries = ConcurrentHashMap<String, DownloadEntry>()
    private val _entriesLiveData = MutableLiveData<DownloadEntry>()
    val entriesLiveData: LiveData<DownloadEntry> get() = _entriesLiveData

    private val _downloadStatus = MutableSharedFlow<DownloadEntry>(replay = 0, extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    val downloadStatus: SharedFlow<DownloadEntry> get() = _downloadStatus

    //SupervisorJob()，保证 databaseScope 的生命周期管理，并避免子协程失败影响其他协程
    private val databaseScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    companion object {
        private const val TAG = "DataChanger"

        @Volatile
        private var instance: DataChanger? = null

        fun getInstance(context: Context): DataChanger {
            return instance ?: synchronized(this) {
                instance ?: DataChanger(context.applicationContext).also { instance = it }
            }
        }
    }

    fun postStatus(entry: DownloadEntry) {

        operatedEntries[entry.key] = entry

        databaseScope.launch {
            val downloadDao = DownloadDatabase.getInstance(context).downloadEntryDao()
            val existingDownload = withContext(Dispatchers.IO) {
                downloadDao.getDownloadById(entry.key)
            }

            existingDownload?.let {
                entry.id = it.id // 从数据库中同步PID(如果存在)
            }

            withContext(Dispatchers.IO) {
                downloadDao.insertOrUpdate(entry)
                Trace.d("postStatus insertOrUpdateentry : $entry")
            }

            if (entry.status == DownloadStatus.COMPLETED) {
                //下载完 删除数据库中的记录
                withContext(Dispatchers.IO) {
                    //由于外部传来的entry 不带pid,所以需要从数据库中查询
                    val realEntry = downloadDao.getDownloadById(entry.key)
                    val delete = realEntry?.let { downloadDao.deleteDownload(it) }

                    Trace.d("postStatus COMPLETED realEntry url : ${realEntry?.url} delete : $delete")
                }
            }

            // 用主线程上的最新条目更新LiveData
            withContext(Dispatchers.Main) {
                _entriesLiveData.value = entry //livedata

                _downloadStatus.emit(entry)//flow
            }
        }
    }

    fun queryAllRecoverableEntries(): List<DownloadEntry> {
        return operatedEntries.values.filter { it.status == DownloadStatus.PAUSED }
    }

    fun queryDownloadEntryById(id: String): DownloadEntry? {
        return operatedEntries[id]
    }

    fun addToOperatedEntryMap(id: String, downloadEntry: DownloadEntry) {
        operatedEntries[id] = downloadEntry
    }

    fun containsDownloadEntry(id: String): Boolean {
        return operatedEntries.containsKey(id)
    }

    fun deleteDownloadEntry(key: String): Boolean {
        databaseScope.launch {
            withContext(Dispatchers.IO) {
                DownloadDatabase.getInstance(context).downloadEntryDao().deleteDownloadByKey(key)
            }
        }
        return operatedEntries.remove(key) != null
    }
}
